sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala

```diff
@@ -518,29 +520,47 @@ class RapidsBufferCatalog(
         // If the store has 0 spillable bytes left, it has exhausted.
         var exhausted = false
 
-        while (!exhausted && !rmmShouldRetryAlloc &&
+        val buffersToFree = new ArrayBuffer[RapidsBuffer]()
+        try {
+          while (!exhausted && !rmmShouldRetryAlloc &&
             store.currentSpillableSize > targetTotalSize) {
-          val mySpillCount = spillCount
-          synchronized {
-            if (spillCount == mySpillCount) {
-              spillCount += 1
-              val nextSpillable = store.nextSpillable()
-              if (nextSpillable != null) {
-                // we have a buffer (nextSpillable) to spill
-                spillAndFreeBuffer(nextSpillable, spillStore, stream)
-                totalSpilled += nextSpillable.getMemoryUsedBytes
+            val mySpillCount = spillCount
+            synchronized {
+              if (spillCount == mySpillCount) {
+                spillCount += 1
+                val nextSpillable = store.nextSpillable()
+                if (nextSpillable != null) {
+                  // we have a buffer (nextSpillable) to spill
+                  // spill it and store it in `buffersToFree` to
+                  // free all in one go after a synchronize.
+                  spillBuffer(nextSpillable, spillStore, stream)
+                    .foreach(buffersToFree.append(_))
+                  totalSpilled += nextSpillable.getMemoryUsedBytes
+                }
+              } else {
+                rmmShouldRetryAlloc = true
               }
-            } else {
-              rmmShouldRetryAlloc = true
             }
-          }
-          if (!rmmShouldRetryAlloc && totalSpilled <= 0) {
-            // we didn't spill in this iteration, exit loop
-            exhausted = true
-            logWarning("Unable to spill enough to meet request. " +
+            if (!rmmShouldRetryAlloc && totalSpilled <= 0) {
+              // we didn't spill in this iteration, exit loop
+              exhausted = true
+              logWarning("Unable to spill enough to meet request. " +
                 s"Total=${store.currentSize} " +
                 s"Spillable=${store.currentSpillableSize} " +
                 s"Target=$targetTotalSize")
+            }
+          }
+        } finally {
+          if (buffersToFree.nonEmpty) {
+            // This is a hack in order to completely synchronize with the GPU before we free
+            // a buffer. It is necessary because of non-synchronous cuDF calls that could fall
+            // behind where the CPU is. Freeing a rapids buffer in these cases needs to wait for
+            // all launched GPU work, otherwise crashes or data corruption could occur.
+            // A more performant implementation would be to synchronize on the thread that created
+            // the buffer via events.
```

Comment:
Nit: IIUC the issue is not with lack of synchronization when the buffer is created, we're already synchronizing on the creation thread's stream when adding the buffer to the spill framework.  Instead this is an issue with a thread reading the buffer asynchronously and another thread tries to free it asynchronously while it's still being read.  That's why synchronizing after the spill copy and not before is still OK.  If it was an issue on creation, we would need to synchronize before the spill copy, not after.
 ```suggestion
            // A more performant implementation would be to synchronize on the thread that read
            // the buffer via events.
```

Reply:
Thanks, that’s a good catch in the comment!